--- pycairo-0.4.0/configure.ac.orig	2005-03-10 09:52:31.000000000 +0100
+++ pycairo-0.4.0/configure.ac	2005-03-28 02:04:08.643151360 +0200
@@ -56,7 +56,7 @@
   with_pygtk=false
 else
   PKG_CHECK_MODULES(CAIRO_GTK,
-    [gtk+-2.0 >= gtk_required_version dnl
+    [gtk+-2.0 >= gtk_required_version
     pygtk-2.0 >= pygtk_required_version],
     [with_pygtk=true], [with_pygtk=false])
   if test -n "$export_dynamic"; then
